src/main/java/org/apache/hadoop/mapred/MesosExecutor.java

```diff
@@ -159,4 +175,82 @@ public void error(ExecutorDriver d, String message) {
   public void shutdown(ExecutorDriver d) {
     LOG.info("Executor asked to shutdown");
   }
+
+  public void revokeSlots() {
+    if (taskTracker == null) {
+      LOG.error("Task tracker is not initialized");
+      return;
+    }
+
+    int maxMapSlots = 0;
+    int maxReduceSlots = 0;
+
+    // TODO(tarnfeld): Sanity check that it's safe for us to change the slots.
+    // Be sure there's nothing running and nothing in the launcher queue.
+
+    // If we expect to have no slots, let's go ahead and terminate the task launchers
+    if (maxMapSlots == 0) {
+      try {
+        Field launcherField = taskTracker.getClass().getDeclaredField("mapLauncher");
+        launcherField.setAccessible(true);
+
+        // Kill the current map task launcher
+        ((TaskTracker.TaskLauncher) launcherField.get(taskTracker)).interrupt();
+      } catch (ReflectiveOperationException e) {
+        LOG.fatal("Failed updating map slots due to error with reflection", e);
+      }
+    }
+
+    if (maxReduceSlots == 0) {
+      try {
+        Field launcherField = taskTracker.getClass().getDeclaredField("reduceLauncher");
+        launcherField.setAccessible(true);
+
+        // Kill the current reduce task launcher
+        ((TaskTracker.TaskLauncher) launcherField.get(taskTracker)).interrupt();
+      } catch (ReflectiveOperationException e) {
+        LOG.fatal("Failed updating reduce slots due to error with reflection", e);
+      }
+    }
+
+    // Configure the new slot counts on the task tracker
+    taskTracker.setMaxMapSlots(maxMapSlots);
+    taskTracker.setMaxReduceSlots(maxReduceSlots);
+
+    // If we have zero slots left, commit suicide when no jobs are running
+    if (maxMapSlots + maxReduceSlots == 0) {
+      scheduleSuicideTimer();
+    }
+  }
+
+  protected void scheduleSuicideTimer() {
+    timerScheduler.schedule(new Runnable() {
+      @Override
+      public void run() {
+        if (taskTracker == null) {
+          return;
+        }
+
+        LOG.info("Checking to see if TaskTracker has no running jobs");
+        int runningJobs = taskTracker.runningJobs.size();
+
+        // Check to see if the number of running jobs on the task tracker is zero
+        if (runningJobs == 0) {
```

Comment:
This check entirely based on `TaskTracker.runningJobs` is not sufficient. It appears this map is not kept 100% up to date especially when failed jobs/tasks are concerned. Perhaps we should look into `TaskTracker.runningTasks` instead.